sdk/iot/azure-iot-modelsrepository/azure/iot/modelsrepository/resolver.py

```diff
@@ -0,0 +1,148 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import six
+import json
+import abc
+import re
+import os
+from .chainable_exception import ChainableException
+
+logger = logging.getLogger(__name__)
+
+
+class ResolverError(ChainableException):
+    pass
+
+
+class DtmiResolver(object):
+    def __init__(self, fetcher):
+        """
+        :param fetcher: A Fetcher configured to an endpoint to resolve DTMIs from
+        :type fetcher: :class:`azure.iot.modelsrepository.resolver`
+        """
+        self.fetcher = fetcher
+
+    def resolve(self, dtmis, expanded_model=False):
+        """Resolve a DTMI from the configured endpoint and return the resulting JSON model.
+
+        :param list[str] dtmis: DTMIs to resolve
+        :param bool expanded_model: Indicates whether to resolve a regular or expanded model
+
+        :raises: ValueError if the DTMI is invalid
+        :raises: ResolverError if the DTMI cannot be resolved to a model
+
+        :returns: A dictionary mapping DTMIs to models
+        :rtype: dict
+        """
+        model_map = {}
+        for dtmi in dtmis:
+            dtdl_path = _convert_dtmi_to_path(dtmi)
+            if expanded_model:
+                dtdl_path = dtdl_path.replace(".json", ".expanded.json")
+
+            try:
+                dtdl = self.fetcher.fetch(dtdl_path)
+            except FetcherError as e:
+                raise ResolverError("Failed to resolve dtmi: {}".format(dtmi), cause=e)
+
+            if expanded_model:
+                for model in dtdl:
+                    model_map[model["@id"]] = model
+            else:
+                model_map[dtmi] = dtdl
```

Comment:
Dtmis are case sensitive. When not grabbing models from the expanded model file, the client needs to ensure the dtmi extracted from a fetched model matches in case with what was requested/expected.

Reply:
It was my understanding as per our previous conversations in meetings that we are not validating any model contents beyond what was strictly necessary in order to find dependencies.

If this is considered necessary, I can make the change, however what is the expected behavior if there is a mismatch?

Reply:
This is necessary. The service convention lowercases all model file paths. The client that fetches models needs to make sure the fetched models definition has exactly the requested dtmi. If there is a mismatch the client needs to raise the idiomatic service exception. In C# we are raising the `RequestFailedException`. At this point in time state of the code, the  ResolverError would fit.